### PR TITLE
Simplify `ConstValue::ScalarPair`

### PR DESCRIPTION
**Pull request by oli-obk**
While looking at #57432 I realized that some of our types for representing constants are very big. This reduces `LazyConst` to 3/4th of its original size and simplifies some code around slices at the same time.

r? @RalfJung

### CODE REVIEW

File: src/librustc/mir/interpret/value.rs [hunk 1]
```diff
@@ -22,22 +22,23 @@ pub enum ConstValue<'tcx> {
     /// Not using the enum `Value` to encode that this must not be `Undef`
     Scalar(Scalar),
 
-    /// Used only for *fat pointers* with layout::abi::ScalarPair
-    ///
-    /// Needed for pattern matching code related to slices and strings.
-    ScalarPair(Scalar, Scalar),
+    /// Used only for slices and strings (`&[T]`, `&str`, `*const [T]`, `*mut str`, `Box<str>`, ...)
+    Slice(Scalar, u64),
```
Comment (varkor) [inline src/librustc/mir/interpret/value.rs]: I like this name change; `ScalarPair` felt a little bit cryptic.

Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]: I guess this cannot be a `Pointer` because of empty slices?

File: src/librustc_mir/const_eval.rs [hunk 2]
```diff
@@ -100,10 +100,7 @@ pub fn op_to_const<'tcx>(
     let normalize = may_normalize
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: The comment above should also be updated: `s/fat pointers/slices/`.

**Comment by RalfJung** [comment]
Nice! r=me with the comment fixed.

File: src/librustc/mir/interpret/value.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc/mir/interpret/value.rs]: Yea. I was looking into it but it would require additional code everywhere or dummy-`AllocId`s.

Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: Okay. This is worth a comment as well.

**Comment by RalfJung** [comment]
Log is empty. Probably spurious.

Cc @rust-lang/infra: same as https://github.com/rust-lang/rust/pull/57351#issuecomment-452981023

**Comment by Centril** [comment]
based on https://github.com/rust-lang/rust/pull/57599#issuecomment-454124204, 
@bors r-

**Comment by oli-obk** [comment]
@bors r=RalfJung

we're allowed to break clippy again

**Comment by Centril** [comment]
Failed in rollup, https://github.com/rust-lang/rust/pull/57911#issuecomment-457795810.

@bors r-

File: src/librustc/mir/interpret/value.rs [hunk 3]
```diff
@@ -22,22 +22,28 @@ pub enum ConstValue<'tcx> {
     /// Not using the enum `Value` to encode that this must not be `Undef`
```
Comment (rasendubi) [inline src/librustc/mir/interpret/value.rs]: The comment above needs an update as there is no more `ScalarPair`
```
/// Represents a constant value in Rust. Scalar and ScalarPair are optimizations which
/// matches the LocalValue optimizations for easy conversions between Value and ConstValue.
```

**Review by rasendubi — COMMENTED** [review]
Also, rustc-guide refers to `ScalarPair` in [`const-eval`][const-eval] and [`miri`][miri] chapters.

[const-eval]: https://github.com/rust-lang/rustc-guide/blob/99e1b1d53656be08654df399fc200584aebb50e4/src/const-eval.md
[miri]: https://github.com/rust-lang/rustc-guide/blob/99e1b1d53656be08654df399fc200584aebb50e4/src/miri.md

**Pull request merged**